Documentation/proposals/202411-zone-aware-sharding.md

```diff
@@ -0,0 +1,295 @@
+
+```text
+0 1 2 | 3 4 5 | 6 7 8 | 9 | shard index
+A B C | A B C | A B C | A | zone
+0 0 0 | 1 1 1 | 2 2 2 | 0 | assignment index
+```
+
+In this case the 2nd assert will warn about double scraping of instances in
+zone A, as the same targets are being assigned to instance 0 and 9.
+
+When `num_shards` is 2 and `len(zones)` is 3 as in `[A..C]`.
+`shards_per_zone` is 1. This yields the following distribution
+
+```text
+0 1   | shard index
+A B C | zone
+0 0 0 | assignment index
+```
+
+In this case the 1st assert will warn about zone C not being scraped.
+
+This second case - a zone not being scraped - should lead to an error in the
+operator, causing the change to not be rolled out, as data would be lost.
+The first case - double scraping - should at minimum cause a warning.
+
+## API changes
+
+Following the algorithm presented above, we suggest the following configuration
+options to be added to the [Prometheus](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#monitoring.coreos.com/v1.Prometheus) and PrometheusAgent custom resource definitions.
+
+All values used in this snippet should also be the defaults for their
+corresponding keys.
+
+```yaml
+spec:
+  shardingStrategy:
+     # Select a sharding mode. Can be 'Classic' or 'Topology'
+    mode: 'Classic'    
+
+    # The following section is only valid if "mode" is set to "Classic"
+    classic:
+        # Metric label used for sharding.
+        sourceLabel: '__address__'
+      
+    # The following section is only valid if "mode" is set to "Topology"
+    # 'topology.kubernetes.io/zone' for 'topology'
+    topology: 
+        # A kubernetes node label defining the topology to be sharded on
+        nodeLabel: 'topology.kubernetes.io/zone'
+        
+        # A prometheus metric containing the topology value of a given target
+        sourceLabel: '__meta_kubernetes_pod_label_topology_kubernetes_io_zone'
+
+        # All topology values to be used by nodeLabel and sourceLabel
+        values: []
+```
+
+The `additionalRelabelConfig` section is meant to allow the `sourceLabel` to be
+generated if needed. This should allow enough flexibility to cover edgecases
+not anticiapted by this proposal.
+
+It is also to be noted that the `topology` section does not use the term `zone`.
+This makes the feature more flexible in case a user needs to shard on e.g.
+regions instead.
+
+## Generated configuration
+
+The following examples are based on the algorithm above.
+Please note that `shard_index` has to be provided by the operator during
+config generation.
+
+We use a replica count of 2 in all examples to illustrate that this value
+does not have any effect, as both replicas will have the same `shared_index`
+assigned.
+
+### classic mode
+
+Given the following configuration
+
+```yaml
+spec:
+  shards: 4
+  replicas: 2
+  shardingStrategy:
+    mode: 'Classic'    
+    classic:
+      sourceLabel: '__address__'
```

Comment:
could this be optional?

Reply:
Yes. I think I put somewhere that `__address__` is just the default value here.
I'll make this more explicit in the API changes section,